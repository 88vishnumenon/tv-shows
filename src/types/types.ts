export interface TVShow {
  showName: string;
  showImage: {medium:string,original:string};
  ratings: string;
  summary: string | TrustedHTML;
  runtime: number;
  genereList:string[];
  id:number;
  officialSite:string;
  status:string;
  schedule: { time: string; days: string[] };

}

export interface TVShowDetails {
  score: string;
  show:TVShowDetailValues
}

export interface TVShowDetailValues{
    averageRuntime: number;
    dvdCountry: string;
    ended: string;
    external: { tvRange: number; thetvdb: number; imdb: string };
    genres: string[];
    id: number;
    image: {medium:string,original:string};
    language: string;
    name: string;
    network: {
      id: number;
      name: string;
      officialSite: string;
      country: { name: string; code: string; timezone: string };
    };
    officialSite: string;
    premiered: string;
    rating: { average: string };
    runTime: number;
    schedule: { time: string; days: string[] };
    status: string;
    summary: string;
    type: string;
    updated: number;
    url: string;
    webChannel: string;
    weight: number;
}
