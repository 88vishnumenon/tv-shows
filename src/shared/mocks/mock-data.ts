import { TVShow } from "../../types/types";

export const mockGenereList: string[] = ["DRAMA", "COMEDY", "THRILLERS"];

export const mockShowList: TVShow[] = [
  {
    showName: "Professor T",
    showImage: {
        "medium": "https://static.tvmaze.com/uploads/images/medium_portrait/482/1205757.jpg",
        "original": "https://static.tvmaze.com/uploads/images/original_untouched/482/1205757.jpg"
      },
    ratings: "7.6",
    summary: "",
    runtime: 60,
    genereList: ["DRAMA", "COMEDY", "THRILLERS"],
    id: 0,
    officialSite: "",
    status: "ENDED",
    schedule: { time: "", days: ["Monday"] },
  },
];





