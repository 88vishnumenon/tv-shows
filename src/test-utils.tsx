import { Provider } from "react-redux";
import { RootState, getStoreWithState, store } from "./store/store";
import { BrowserRouter as Router } from "react-router-dom";
import { render } from "@testing-library/react";
import * as React from "react";
import { TVShow } from "./types/types";

export function renderWithContext(
  element: React.ReactElement,
  state?: RootState
) {
  const store = getStoreWithState(state);
  const utils = render(
    <Provider store={store}>
      <Router>{element}</Router>
    </Provider>
  );
  return { store, ...utils };
}

export function getStateWithItems(
  tvShows: TVShow[]
): RootState {
  const state: RootState = {
    tvshow: {
      showList: tvShows,
      generesList: ["DRAMA"],
      searchString: "",
      error: false,
    },
  };

  return state;
}
