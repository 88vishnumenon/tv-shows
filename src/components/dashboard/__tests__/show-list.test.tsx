import userEvent from "@testing-library/user-event";
import { mockShowList } from "../../../shared/mocks/mock-data";
import { renderWithContext } from "../../../test-utils";
import ShowList from "../show-list";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import * as React from "react";
import { TVShow } from "../../../types/types";

const genereName = "DRAMA";
const mockUsedNavigate = jest.fn();
const mockUsedRoutes = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockUsedNavigate,
  Routes: () => mockUsedRoutes,
}));

function renderShowListComponent(genereName: string, tvShowList: TVShow[]) {
  return renderWithContext(
    <ShowList genereName={genereName} tvShowsList={mockShowList}></ShowList>
  );
}
describe("show-list component", () => {
  it("snapshot is rendered", () => {
    const { asFragment } = renderWithContext(
      <ShowList genereName={genereName} tvShowsList={mockShowList}></ShowList>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("is genere name displayed", () => {
    renderShowListComponent(genereName, mockShowList);
    const genereElement = screen.getByTestId("genere-name");
    expect(genereElement).toHaveTextContent(genereName);
  });

  it("is tv shows  displayed under genere ", () => {
    renderShowListComponent(genereName, mockShowList);
    const cardsElements = screen.getAllByTestId("show-cards");
    expect(cardsElements).toHaveLength(1);
  });

  it("should click on cards navigate to details page", async () => {
    renderShowListComponent(genereName, mockShowList);
    const cardsElement = screen.getAllByTestId("show-cards")[0];
    userEvent.click(cardsElement);
    await waitFor(() => expect(mockUsedNavigate).toHaveBeenCalled(), {
      timeout: 500,
    });
  });

  it("is tv shows  not displayed when genere is not available ", () => {
    renderShowListComponent("HORROR", mockShowList);
    const cardsElements = screen.queryByTestId("show-cards");
    expect(cardsElements).toBe(null);
  });
});
