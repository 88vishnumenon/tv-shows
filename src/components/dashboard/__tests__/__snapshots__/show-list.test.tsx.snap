// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show-list component snapshot is rendered 1`] = `
<DocumentFragment>
  <article
    class="showListrows"
  >
    <div
      class="genereName"
      data-testid="genere-name"
    >
      DRAMA
    </div>
    <section
      class="showWrapper"
    >
      <div
        class="showItem"
        data-testid="show-cards"
      >
        <img
          class="showImg"
          src="https://static.tvmaze.com/uploads/images/medium_portrait/482/1205757.jpg"
        />
        <div
          class="showDetails"
        >
          <span
            class="title"
          >
            Professor T
          </span>
          <section
            class="showDetailsValues"
          >
            <div
              class="favouriteSection"
            >
              <span
                class="fa fa-star checked"
              />
              <span>
                 7.6
              </span>
            </div>
            <div>
              <span
                class="genereList genereLeft"
              >
                DRAMA | COMEDY | THRILLERS
              </span>
            </div>
          </section>
        </div>
      </div>
    </section>
  </article>
</DocumentFragment>
`;
